File: tools/symbol_analyzer.ts
#!/usr/bin/env node

/**
 * Symbol Analyzer Tool - TypeScript Wrapper (Refactored for OpenCode)
 * Extracts and analyzes symbols from binaries via FastAPI server.
 * 
 * This wrapper communicates with the FastAPI job server (http://tools:8000) instead
 * of spawning child processes. It sends HTTP requests and parses responses.
 * 
 * NOTE: Requires /symbol_analyzer endpoint on FastAPI server.
 */

interface SymbolAnalyzerConfig {
  binary_path: string;
  symbol_types?: string;
  demangle?: boolean;
}

interface ToolResult {
  status: 'success' | 'error';
  tool_name: string;
  execution_time_ms: number;
  data: any;
  summary: string;
  metadata: {
    ai_next_step?: string;
    [key: string]: any;
  };
}

async function runSymbolAnalyzer(config: SymbolAnalyzerConfig): Promise<ToolResult> {
  const startTime = Date.now();

  try {
    if (!config.binary_path) {
      throw new Error('binary_path is required');
    }

    // Build request payload
    const payload: Record<string, any> = {
      binary: config.binary_path
    };

    if (config.symbol_types) {
      payload.symbol_types = config.symbol_types;
    }

    if (config.demangle !== undefined) {
      payload.demangle = config.demangle;
    }

    // Call FastAPI server at http://tools:8000/symbol_analyzer
    const apiUrl = 'http://tools:8000/symbol_analyzer';
    let response;
    try {
      response = await fetch(apiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(60000)
      });
    } catch (fetchError) {
      const errorMsg = fetchError instanceof Error ? fetchError.message : String(fetchError);
      const executionTime = Date.now() - startTime;
      return {
        status: 'error',
        tool_name: 'symbol_analyzer',
        execution_time_ms: executionTime,
        data: { error: errorMsg },
        summary: `Failed to connect to FastAPI server at ${apiUrl}: ${errorMsg}`,
        metadata: {
          ai_next_step: 'Ensure the tools container is running on http://tools:8000'
        }
      };
    }

    if (!response.ok) {
      let errorDetail = '';
      try {
        const errorBody = await response.json();
        errorDetail = errorBody.detail || JSON.stringify(errorBody);
      } catch {
        errorDetail = await response.text();
      }

      const executionTime = Date.now() - startTime;
      return {
        status: 'error',
        tool_name: 'symbol_analyzer',
        execution_time_ms: executionTime,
        data: { error: `HTTP ${response.status}`, details: errorDetail },
        summary: `FastAPI server returned error: ${response.status} - ${errorDetail}`,
        metadata: {
          ai_next_step: 'Verify binary file exists and is in a supported format'
        }
      };
    }

    let responseData;
    try {
      responseData = await response.json();
    } catch {
      const executionTime = Date.now() - startTime;
      const rawText = await response.text();
      return {
        status: 'error',
        tool_name: 'symbol_analyzer',
        execution_time_ms: executionTime,
        data: { raw: rawText },
        summary: 'Failed to parse FastAPI response as JSON',
        metadata: {
          ai_next_step: 'Check server logs for symbol extraction errors'
        }
      };
    }

    const executionTime = Date.now() - startTime;

    // Build successful result
    return {
      status: 'success',
      tool_name: 'symbol_analyzer',
      execution_time_ms: executionTime,
      data: responseData,
      summary: `Successfully extracted symbols from ${config.binary_path} (${responseData.count || '?'} symbols found)`,
      metadata: {
        ai_next_step: 'Review symbols for changed functions; correlate with BinDiff and disassembly results'
      }
    };
  } catch (error) {
    const executionTime = Date.now() - startTime;
    const message = error instanceof Error ? error.message : String(error);

    return {
      status: 'error',
      tool_name: 'symbol_analyzer',
      execution_time_ms: executionTime,
      data: { error: message },
      summary: `Symbol analyzer error: ${message}`,
      metadata: {
        ai_next_step: 'Check configuration and ensure binary file is valid'
      }
    };
  }
}

// CLI interface for direct invocation
async function main() {
  const args = process.argv.slice(2);
  const config: SymbolAnalyzerConfig = {
    binary_path: ''
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--binary':
        config.binary_path = args[++i];
        break;
      case '--symbol-types':
        config.symbol_types = args[++i];
        break;
      case '--demangle':
        config.demangle = true;
        break;
    }
  }

  const result = await runSymbolAnalyzer(config);
  console.log(JSON.stringify(result, null, 2));
}

if (require.main === module) {
  main().catch((err) => {
    console.error(JSON.stringify({
      status: 'error',
      tool_name: 'symbol_analyzer',
      execution_time_ms: 0,
      data: { error: err.message },
      summary: `Fatal error: ${err.message}`,
      metadata: { ai_next_step: 'Check logs' }
    }, null, 2));
    process.exit(1);
  });
}

export { runSymbolAnalyzer, SymbolAnalyzerConfig, ToolResult };


async function main() {
  const args = process.argv.slice(2);
  const config: SymbolAnalyzerConfig = {
    binary_path: '',
    demangle: true,
    output_format: 'json'
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--binary': config.binary_path = args[++i]; break;
      case '--symbol-types': config.symbol_types = args[++i]; break;
      case '--demangle': config.demangle = true; break;
      case '--output-format': config.output_format = args[++i] as 'json' | 'text'; break;
    }
  }

  const result = await runSymbolAnalyzer(config);
  console.log(JSON.stringify(result, null, 2));
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: tools/symbol_analyzer.py
#!/usr/bin/env python3

"""
Symbol Analyzer Tool - Python Implementation
Extracts and analyzes symbols using `nm` and `objdump` commands
"""

import json
import sys
import os
import argparse
import subprocess
import re
from pathlib import Path
from typing import Dict, List, Any, Tuple

class SymbolAnalyzer:
    def __init__(self, binary_path: str, symbol_types: str = '',
                 demangle: bool = True):
        self.binary_path = binary_path
        self.symbol_types = symbol_types
        self.demangle = demangle

    def extract_nm_symbols(self) -> List[Dict[str, str]]:
        """Extract symbols using the `nm` command"""
        try:
            cmd = ['nm']

            if self.demangle:
                cmd.append('-C')

            cmd.extend(['-S', self.binary_path])

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.returncode != 0:
                raise RuntimeError(f"nm failed: {result.stderr.strip()}")

            symbols = []
            type_chars = {'T','t','D','d','B','b','U','u','W','w','A','a','C','c',
                          'G','g','S','s','I','i','N','n','O','o','P','p','R','r',
                          'V','v','Z','z'}

            for line in result.stdout.strip().split('\n'):
                if not line.strip():
                    continue
                parts = line.split()
                # nm -C -S output patterns:
                # 4 parts: address, size, type, name (defined with size)
                # 3 parts: address, type, name (defined without size)
                # 2 parts: type, name (undefined, no address)
                if len(parts) >= 4:
                    symbols.append({
                        'address': parts[0],
                        'size': parts[1],
                        'type': parts[2],
                        'name': ' '.join(parts[3:])
                    })
                elif len(parts) == 3 and parts[1] in type_chars:
                    symbols.append({
                        'address': parts[0],
                        'size': '',
                        'type': parts[1],
                        'name': parts[2]
                    })
                elif len(parts) >= 2:
                    symbols.append({
                        'address': '',
                        'size': '',
                        'type': parts[0],
                        'name': ' '.join(parts[1:])
                    })

            return symbols

        except Exception as e:
            raise RuntimeError(f"nm extraction failed: {str(e)}")

    def extract_objdump_symbols(self) -> List[Dict[str, Any]]:
        """Extract symbols using `objdump -t`"""
        try:
            cmd = ['objdump', '-t']

            if self.demangle:
                cmd.append('-C')

            cmd.append(self.binary_path)

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.returncode != 0:
                return []

            sections = []
            for line in result.stdout.strip().split('\n'):
                if not line.strip() or line.startswith('SYMBOL TABLE:'):
                    continue

                match = re.match(
                    r'^([0-9a-fA-F]+)\s+([lg\s])\s+([\w\s]+?)\s+([0-9a-fA-F]+)\s+(.+)$',
                    line
                )
                if match:
                    sections.append({
                        'value': match.group(1),
                        'flags': match.group(2).strip(),
                        'type': match.group(3).strip(),
                        'size': match.group(4),
                        'name': match.group(5).strip()
                    })

            return sections

        except Exception as e:
            return []

    def analyze(self) -> Dict[str, Any]:
        """Analyze symbol tables"""
        try:
            if not Path(self.binary_path).exists():
                raise FileNotFoundError(f"Binary not found: {self.binary_path}")

            # Get symbols from nm
            nm_symbols = self.extract_nm_symbols()
            objdump_sections = self.extract_objdump_symbols()

            # Categorize symbols by type
            defined_functions = [s for s in nm_symbols if s['type'] in ('T', 't', 'W', 'w')]
            undefined_functions = [s for s in nm_symbols if s['type'] == 'U']
            data_symbols = [s for s in nm_symbols if s['type'] in ('D', 'd', 'B', 'b', 'G', 'g')]

            # Detect security-relevant symbols
            security_patterns = {
                'string_functions': ['strcpy', 'strncpy', 'strcat', 'strncat', 'strlen', 'strcmp',
                                     'strncmp', 'strdup', 'strndup', 'memcpy', 'memmove', 'memset',
                                     'memcmp'],
                'format_functions': ['printf', 'sprintf', 'snprintf', 'fprintf', 'scanf', 'sscanf'],
                'memory_functions': ['malloc', 'calloc', 'realloc', 'free', 'mmap', 'munmap'],
                'io_functions': ['read', 'write', 'open', 'close', 'fread', 'fwrite', 'fopen', 'fclose',
                                 'send', 'recv', 'accept', 'bind', 'listen', 'connect'],
                'auth_functions': ['login', 'authenticate', 'authorize', 'verify', 'check_password',
                                   'validate_user', 'encrypt', 'decrypt', 'hash']
            }

            security_symbols = {}
            for category, patterns in security_patterns.items():
                matches = [
                    s['name'] for s in nm_symbols
                    if any(p in s['name'] for p in patterns)
                ]
                if matches:
                    security_symbols[category] = matches[:20]

            # Filter by requested types
            if self.symbol_types:
                type_filter = set(self.symbol_types.split(','))
                nm_symbols = [s for s in nm_symbols if s['type'] in type_filter]

            return {
                'total_symbols': len(nm_symbols),
                'defined_functions': len(defined_functions),
                'undefined_functions': len(undefined_functions),
                'data_symbols': len(data_symbols),
                'function_names': [s['name'] for s in defined_functions[:200]],
                'imported_functions': [s['name'] for s in undefined_functions[:200]],
                'security_relevant': security_symbols,
                'sections': objdump_sections[:50],
                'sample_symbols': nm_symbols[:100],
                'summary': f'Found {len(nm_symbols)} symbols ({len(defined_functions)} defined functions, {len(undefined_functions)} imports) in {self.binary_path}'
            }

        except Exception as e:
            return {'error': str(e), 'summary': f'Symbol analysis failed: {str(e)}'}


def main():
    parser = argparse.ArgumentParser(description='Symbol Analyzer Tool')
    parser.add_argument('--binary', required=True, help='Path to binary file')
    parser.add_argument('--symbol-types', help='Filter by symbol types (comma-separated, e.g., T,t,U)')
    parser.add_argument('--demangle', action='store_true', default=True, help='Demangle C++ symbols')
    parser.add_argument('--output-format', choices=['json', 'text'], default='json')

    args = parser.parse_args()

    try:
        analyzer = SymbolAnalyzer(
            binary_path=args.binary,
            symbol_types=args.symbol_types or '',
            demangle=args.demangle
        )

        analysis_result = analyzer.analyze()

        output = {
            'data': analysis_result,
            'summary': analysis_result.get('summary', 'Symbol analysis completed'),
            'metadata': {
                'binary': args.binary,
                'symbol_types_filter': args.symbol_types,
                'demangle': args.demangle,
                'ai_next_step': 'Cross-reference symbol changes with BinDiff results to identify security-relevant functions'
            }
        }

        if args.output_format == 'json':
            print(json.dumps(output, indent=2))
        else:
            print(f"Symbol Analyzer Report")
            print(f"======================")
            print(f"Binary: {args.binary}")
            print(f"Total Symbols: {analysis_result.get('total_symbols', 0)}")
            print(f"Defined Functions: {analysis_result.get('defined_functions', 0)}")
            print(f"Undefined Functions (Imports): {analysis_result.get('undefined_functions', 0)}")
            print(f"Data Symbols: {analysis_result.get('data_symbols', 0)}")
            if analysis_result.get('security_relevant'):
                for cat, syms in analysis_result['security_relevant'].items():
                    print(f"  {cat}: {', '.join(syms[:5])}")

    except Exception as e:
        error_output = {
            'error': str(e),
            'summary': f'Symbol analysis failed: {str(e)}',
            'metadata': {'ai_next_step': 'Check binary and nm/objdump availability'}
        }
        print(json.dumps(error_output, indent=2))
        sys.exit(1)


if __name__ == '__main__':
    main()
